wordpress-mobile/WordPress-Android
# Refactor 'showCancelMediaUploadDialog' to Prevent Dismissal on Screen Rotation

## Description

Partial fix for https://github.com/wordpress-mobile/WordPress-Android/issues/14676

The `showCancelMediaUploadDialog`, which is shown when tapping an image that's still in the process of uploading, is currently dismissed/lost with a screen rotation. 

This can lead to buggy behavior, as shown in the following screenshots, where the image will permanently look like it's still uploading, with any attempts to tap on it resulting in the dialog appearing again, even though the image has been uploaded to the site's media library:

| Before Rotation  | After Rotation |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/118637471-bf43c380-b7cd-11eb-9cd6-2dfea96e1b11.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/118637534-d2ef2a00-b7cd-11eb-97fb-9e8dafdc00e2.png" width="100%"> |

This PR seeks to fix the above issue by refactoring the dialog so that it survives screen rotation. A custom dialog fragment, `GutenbergDialogFragment`, is used to achieve this.

### GutenbergDialogFragment

A [new `GutenbergDialogFragment` class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt) was introduced in https://github.com/wordpress-mobile/WordPress-Android/pull/14503. This class is based on [the `BasicFragmentDialog` class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt), with some key differences being the removal of the third `setNeutralButton`, [an addition of a mediaId](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R39) to the constructor, and the ability for [the parent fragment to implement the positive/negative button interfaces](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R86). 

Part of the goal behind creating the new `GutenbergDialogFragment` class was to allow for easier refactoring of the other dialogs listed in https://github.com/wordpress-mobile/WordPress-Android/issues/14676, which are all currently lost/dismissed with a screen rotation. With that goal in mind, [the mediaId param is refactored with this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/572e9b3b5ff204edf2d2800466e2bc579c30b02b) to utilize generics and accept any data type. Its [message param has also been made optional](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/b0cefc8938f514455882588406edef11b5c1c72b), as not all dialogs declared in `GutenbergEditorFragment` include a message. 

The aforementioned changes make it relatively straightforward to [refactor `showCancelMediaUploadDialog` to use the custom dialog fragment](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/84a7b12d48dbfad9bcebe3ca6c394d362d952721).

## Testing Steps

* Navigate to the post editor.
* Add a new image block and choose the option to upload an image directly from your device.
* Immediately tap on the image block as the upload begins in order to generate the "cancel" dialog.
* Switch the orientation of the device and confirm that the dialog is still in place.
* Tap the "no" or "yes" buttons after rotating the screen to confirm that they still work as expected.
* Try a few combinations of rotations and button pressing to confirm there are no unintended or buggy side effects to this recent change.

## Regression Notes
1. Potential unintended areas of impact

The main unintended side effect of this PR would be for the dialog's existing "yes" or "no" flows to break.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I following the outlined testing steps, with various combinations of rotations and button pressing to see if I could create any problems or issues.

3. What automated tests I added (or what prevented me from doing so)

Automated tests didn't feel warranted given the size of this change.

PR submission checklist:
